File: phase3_gold/mcp_servers/mcp_server_email.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Email MCP Server - Gold Tier
============================
Model Context Protocol (MCP) server for sending emails via Gmail

Port: 3003
Endpoints:
  POST /send-email    - Send email
  POST /send-briefing - Send CEO briefing via email
  GET  /health        - Health check
"""

import os
import sys
import json
import smtplib
import io
from datetime import datetime
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse
from email.mime.text import MIMEText
from email.mime.multipart import MIMEMultipart
from email.mime.base import MIMEBase
from email import encoders
from pathlib import Path

# Fix Windows console encoding
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8')

# Add parent directory to path
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

try:
    from dotenv import load_dotenv
    load_dotenv()
except:
    pass

PORT = 3003
PHASE3_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# Gmail configuration
GMAIL_EMAIL = os.getenv("GMAIL_EMAIL", "")
GMAIL_PASSWORD = os.getenv("GMAIL_PASSWORD", "")  # App password
SMTP_SERVER = "smtp.gmail.com"
SMTP_PORT = 587


class EmailMCPServer(BaseHTTPRequestHandler):
    """HTTP Request Handler for Email MCP Server"""
    
    stats = {
        "requests": 0,
        "emails_sent": 0,
        "emails_failed": 0,
        "start_time": datetime.now().isoformat()
    }
    
    def log_message(self, format, *args):
        print(f"[{datetime.now().strftime('%Y-%m-%d %H:%M:%S')}] {args[0]}")
    
    def send_json_response(self, data, status=200):
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.end_headers()
        self.wfile.write(json.dumps(data, indent=2).encode())
    
    def send_error_response(self, message, status=500):
        self.send_json_response({
            "success": False,
            "error": message
        }, status)
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
    
    def do_GET(self):
        EmailMCPServer.stats["requests"] += 1
        parsed_path = urlparse(self.path)
        path = parsed_path.path
        
        if path == '/health':
            self.handle_health()
        elif path == '/stats':
            self.handle_stats()
        else:
            self.send_error_response("Not found", 404)
    
    def do_POST(self):
        EmailMCPServer.stats["requests"] += 1
        parsed_path = urlparse(self.path)
        path = parsed_path.path
        
        if path == '/send-email':
            self.handle_send_email()
        elif path == '/send-briefing':
            self.handle_send_briefing()
        else:
            self.send_error_response("Not found", 404)
    
    def handle_health(self):
        """Health check endpoint"""
        self.send_json_response({
            "status": "healthy",
            "service": "email-mcp-server",
            "port": PORT,
            "credentials_set": bool(GMAIL_EMAIL and GMAIL_PASSWORD),
            "stats": EmailMCPServer.stats
        })
    
    def handle_stats(self):
        """Server statistics"""
        self.send_json_response({
            "service": "email-mcp-server",
            "stats": EmailMCPServer.stats
        })
    
    def handle_send_email(self):
        """Send email endpoint"""
        try:
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length).decode('utf-8')
            data = json.loads(body)
            
            to_email = data.get('to')
            subject = data.get('subject')
            body_text = data.get('body')
            cc = data.get('cc')
            html = data.get('html', False)
            
            if not to_email or not subject or not body_text:
                self.send_error_response("Missing required fields: to, subject, body", 400)
                return
            
            result = self.send_email(to_email, subject, body_text, cc, html)
            
            if result['success']:
                EmailMCPServer.stats["emails_sent"] += 1
            else:
                EmailMCPServer.stats["emails_failed"] += 1
            
            self.send_json_response(result)
            
        except json.JSONDecodeError:
            self.send_error_response("Invalid JSON", 400)
        except Exception as e:
            EmailMCPServer.stats["emails_failed"] += 1
            self.send_error_response(str(e))
    
    def handle_send_briefing(self):
        """Send CEO briefing via email"""
        try:
            content_length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(content_length).decode('utf-8')
            data = json.loads(body)
            
            to_email = data.get('to')
            
            if not to_email:
                self.send_error_response("Missing 'to' email", 400)
                return
            
            # Read CEO briefing
            briefing_file = os.path.join(PHASE3_DIR, "CEO_Briefing.md")
            if not os.path.exists(briefing_file):
                self.send_error_response("CEO_Briefing.md not found", 404)
                return
            
            with open(briefing_file, 'r', encoding='utf-8') as f:
                briefing_content = f.read()
            
            result = self.send_email(
                to=to_email,
                subject=f"CEO Weekly Briefing - {datetime.now().strftime('%Y-%m-%d')}",
                body=briefing_content,
                html=False
            )
            
            if result['success']:
                EmailMCPServer.stats["emails_sent"] += 1
            
            self.send_json_response(result)
            
        except Exception as e:
            EmailMCPServer.stats["emails_failed"] += 1
            self.send_error_response(str(e))
    
    def send_email(self, to: str, subject: str, body: str, cc: str = None, html: bool = False) -> dict:
        """Send an email with optional CC"""
        
        if not GMAIL_EMAIL or not GMAIL_PASSWORD:
            return {
                'success': False,
                'message': 'Gmail credentials not configured. Set GMAIL_EMAIL and GMAIL_PASSWORD in .env'
            }
        
        try:
            msg = MIMEMultipart()
            msg['From'] = GMAIL_EMAIL
            msg['To'] = to
            msg['Subject'] = subject
            
            if cc:
                msg['Cc'] = cc
            
            msg_type = 'html' if html else 'plain'
            msg.attach(MIMEText(body, msg_type))
            
            # Connect to SMTP server
            server = smtplib.SMTP(SMTP_SERVER, SMTP_PORT)
            server.starttls()
            server.login(GMAIL_EMAIL, GMAIL_PASSWORD)
            
            # Get all recipient addresses
            recipients = [to]
            if cc:
                recipients.extend(cc.split(','))
            
            server.sendmail(GMAIL_EMAIL, recipients, msg.as_string())
            server.quit()
            
            return {
                'success': True,
                'message': f'Email sent to {to}',
                'timestamp': datetime.now().isoformat()
            }
            
        except smtplib.SMTPAuthenticationError:
            return {
                'success': False,
                'message': 'SMTP Authentication failed. Check Gmail App Password.'
            }
        except Exception as e:
            return {
                'success': False,
                'message': f'Failed to send email: {str(e)}'
            }


def run_server():
    """Run the Email MCP Server"""
    server_address = ('', PORT)
    httpd = HTTPServer(server_address, EmailMCPServer)
    
    print(f"📧 Email MCP Server running on port {PORT}")
    print(f"   Endpoints:")
    print(f"   POST /send-email     - Send email")
    print(f"   POST /send-briefing  - Send CEO briefing")
    print(f"   GET  /health         - Health check")
    print(f"   GET  /stats          - Server statistics")
    print(f"\n   Gmail: {GMAIL_EMAIL if GMAIL_EMAIL else 'NOT CONFIGURED'}")
    print(f"   Started at: {EmailMCPServer.stats['start_time']}")
    
    httpd.serve_forever()


if __name__ == '__main__':
    try:
        run_server()
    except KeyboardInterrupt:
        print("\n\nEmail MCP Server stopped")
    except Exception as e:
        print(f"Server error: {e}")

File: phase3_gold/mcp_servers/mcp_server_email.js
#!/usr/bin/env node
/**
 * Email MCP Server - Gold Tier
 * ============================
 * Send emails with attachments via Gmail
 * 
 * Port: 3003
 * Endpoints:
 *   POST /send-email      - Send email
 *   POST /send-invoice    - Send invoice with attachment
 *   GET  /health          - Health check
 *   GET  /stats           - Server statistics
 */

import express from 'express';
import cors from 'cors';
import nodemailer from 'nodemailer';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const app = express();
const PORT = 3003;

// Middleware
app.use(cors());
app.use(express.json({ limit: '10mb' }));

// Load environment variables
import dotenv from 'dotenv';
dotenv.config({ path: path.join(__dirname, '..', '.env') });

// Gmail configuration
const GMAIL_EMAIL = process.env.GMAIL_EMAIL || '';
const GMAIL_PASSWORD = process.env.GMAIL_PASSWORD || '';

// Create transporter
const createTransporter = () => {
  return nodemailer.createTransport({
    service: 'gmail',
    auth: {
      user: GMAIL_EMAIL,
      pass: GMAIL_PASSWORD
    }
  });
};

// Store server stats
const stats = {
  requests: 0,
  emails_sent: 0,
  emails_failed: 0,
  invoices_sent: 0,
  startTime: new Date().toISOString()
};

// Logs directory
const LOGS_DIR = path.join(__dirname, '..', 'Logs');
if (!fs.existsSync(LOGS_DIR)) {
  fs.mkdirSync(LOGS_DIR, { recursive: true });
}

// Health check endpoint
app.get('/health', (req, res) => {
  stats.requests++;
  res.json({
    status: 'healthy',
    service: 'email-mcp-server',
    port: PORT,
    credentials_set: !!(GMAIL_EMAIL && GMAIL_PASSWORD),
    uptime: Date.now() - new Date(stats.startTime).getTime(),
    stats: stats
  });
});

// Get stats endpoint
app.get('/stats', (req, res) => {
  stats.requests++;
  res.json({
    service: 'email-mcp-server',
    stats: stats
  });
});

// Send email endpoint
app.post('/send-email', async (req, res) => {
  stats.requests++;
  
  const { to, subject, body, html = false, attachment } = req.body;
  
  // Validation
  if (!to || !subject || !body) {
    return res.status(400).json({ 
      success: false, 
      error: 'Missing required fields: to, subject, body' 
    });
  }
  
  if (!GMAIL_EMAIL || !GMAIL_PASSWORD) {
    return res.status(500).json({ 
      success: false, 
      error: 'Gmail credentials not configured. Set GMAIL_EMAIL and GMAIL_PASSWORD in .env' 
    });
  }
  
  try {
    const transporter = createTransporter();
    
    // Prepare email options
    const mailOptions = {
      from: GMAIL_EMAIL,
      to: to,
      subject: subject,
      text: html ? undefined : body,
      html: html ? body : undefined
    };
    
    // Add attachment if provided
    if (attachment) {
      if (fs.existsSync(attachment)) {
        mailOptions.attachments = [
          {
            filename: path.basename(attachment),
            path: attachment
          }
        ];
        console.log(`Attachment added: ${attachment}`);
      } else {
        console.warn(`Attachment not found: ${attachment}`);
      }
    }
    
    // Send email
    const info = await transporter.sendMail(mailOptions);
    
    stats.emails_sent++;
    
    // Log transaction
    logTransaction('email', { to, subject, attachment, messageId: info.messageId });
    
    res.json({
      success: true,
      message: 'Email sent successfully',
      messageId: info.messageId,
      timestamp: new Date().toISOString()
    });
    
  } catch (error) {
    stats.emails_failed++;
    console.error('Send email error:', error);
    
    res.status(500).json({
      success: false,
      error: error.message || 'Failed to send email'
    });
  }
});

// Send invoice endpoint (specialized for invoices)
app.post('/send-invoice', async (req, res) => {
  stats.requests++;
  
  const { 
    to, 
    invoice_number, 
    invoice_path, 
    amount,
    client_name 
  } = req.body;
  
  // Validation
  if (!to || !invoice_number || !invoice_path) {
    return res.status(400).json({ 
      success: false, 
      error: 'Missing required fields: to, invoice_number, invoice_path' 
    });
  }
  
  if (!GMAIL_EMAIL || !GMAIL_PASSWORD) {
    return res.status(500).json({ 
      success: false, 
      error: 'Gmail credentials not configured' 
    });
  }
  
  try {
    const transporter = createTransporter();
    
    // Create professional email body
    const emailBody = `
Dear ${client_name || 'Valued Client'},

Please find attached your invoice ${invoice_number} for the amount of $${amount}.

Payment is due within 30 days of receipt.

If you have any questions, please don't hesitate to contact us.

Thank you for your business!

Best regards,
AI Employee Services
${GMAIL_EMAIL}
    `.trim();
    
    // Prepare email options
    const mailOptions = {
      from: GMAIL_EMAIL,
      to: to,
      subject: `Invoice ${invoice_number} - $${amount}`,
      text: emailBody,
      attachments: []
    };
    
    // Add invoice attachment
    if (fs.existsSync(invoice_path)) {
      mailOptions.attachments.push({
        filename: path.basename(invoice_path),
        path: invoice_path
      });
    } else {
      console.warn(`Invoice file not found: ${invoice_path}`);
    }
    
    // Send email
    const info = await transporter.sendMail(mailOptions);
    
    stats.invoices_sent++;
    
    // Log transaction
    logTransaction('invoice', { 
      to, 
      invoice_number, 
      amount, 
      invoice_path,
      messageId: info.messageId 
    });
    
    res.json({
      success: true,
      message: 'Invoice sent successfully',
      messageId: info.messageId,
      timestamp: new Date().toISOString()
    });
    
  } catch (error) {
    stats.emails_failed++;
    console.error('Send invoice error:', error);
    
    res.status(500).json({
      success: false,
      error: error.message || 'Failed to send invoice'
    });
  }
});

// Helper function to log transactions
function logTransaction(type, data) {
  const timestamp = new Date().toISOString();
  const logEntry = {
    timestamp,
    type,
    ...data
  };
  
  const logFile = path.join(LOGS_DIR, 'email_mcp_transactions.jsonl');
  fs.appendFileSync(logFile, JSON.stringify(logEntry) + '\n');
  
  // Also log to daily file
  const dateStr = new Date().toISOString().split('T')[0];
  const dailyLogFile = path.join(LOGS_DIR, `${dateStr}.json`);
  
  let dailyLogs = [];
  if (fs.existsSync(dailyLogFile)) {
    try {
      dailyLogs = JSON.parse(fs.readFileSync(dailyLogFile, 'utf-8'));
    } catch (e) {
      dailyLogs = [];
    }
  }
  
  dailyLogs.push(logEntry);
  fs.writeFileSync(dailyLogFile, JSON.stringify(dailyLogs, null, 2));
  
  console.log(`Transaction logged: ${type} to ${data.to}`);
}

// Start server
app.listen(PORT, () => {
  console.log('📧 Email MCP Server running on port', PORT);
  console.log('   Endpoints:');
  console.log('   POST /send-email       - Send email');
  console.log('   POST /send-invoice     - Send invoice with attachment');
  console.log('   GET  /health           - Health check');
  console.log('   GET  /stats            - Server statistics');
  console.log('');
  console.log('   Gmail:', GMAIL_EMAIL || 'NOT CONFIGURED');
  console.log('   Started at:', stats.startTime);
  console.log('');
  console.log('Example usage:');
  console.log('   curl -X POST http://localhost:3003/send-email \\');
  console.log('     -H "Content-Type: application/json" \\');
  console.log('     -d \'{"to":"client@example.com","subject":"Test","body":"Hello"}\'');
});
